[dns-client] add support for arbitrary DNS record type query

abtink opened the pull request:
This commit introduces a new feature in `Dns::Client` to support sending DNS queries for arbitrary record types. Callers are notified of received response via a callback. New methods are provided to parse and read all records in the response. Public APIs and related CLI commands for this new feature are also added.

The `OPENTHREAD_CONFIG_DNS_CLIENT_ARBITRARY_RECORD_QUERY_ENABLE` build configuration is added to control this feature. This allows projects that do not require arbitrary DNS query functionality to disable it, avoiding its associated firmware code size overhead.

Importantly, if a retrieved record type is PTR, CNAME, DNAME, NS, or SRV, the record data in the received response contains a DNS name which may use DNS name compression. For these specific record types, the record data is first decompressed such that it contains the uncompressed DNS name. For all other record types, the record data is read and provided as it appears in the received response message.

----

While this PR/commit itself doesn't contain any new tests for the new DNS client API, the PR #11357 (which adds support for this on the DNSSD name server/resolver (SRP and discovery proxy)) includes detailed tests that also use and cover the newly added behavior in this PR.

Vyrastas commented on src/cli/README.md:
```diff
@@ -1564,6 +1564,14 @@ Service instance label is provided first, followed by the service name (note tha
+
+Sends a DNS query for a given record type and DNS name. DNS name is provided as a first label, followed by the next labels which are dot '.' separated. Note that the first label can itself contain the dot '.' character.
+
+The parameters after `next-labels` are optional. Any unspecified (or zero) value for these optional parameters is eplaced by the value from the current default config (`dns config`).
```
```suggestion
The parameters after `next-labels` are optional. Any unspecified (or zero) value for these optional parameters is replaced by the value from the current default config (`dns config`).
```

Vyrastas commented on src/cli/cli_dns.cpp:
```diff
@@ -410,6 +410,48 @@ otError Dns::ProcessService(Arg aArgs[], ResolveServiceFn aResolveServiceFn)
+
+/**
+ * @cli dns query
+ * @cparam dns query @ca{record-type} @ca{first-label} @ca{next-labels} <!--
```
Can we add an example for this?

abtink replied:
Added example.

Vyrastas commented on src/cli/cli_dns.cpp:
```diff
@@ -410,6 +410,55 @@ otError Dns::ProcessService(Arg aArgs[], ResolveServiceFn aResolveServiceFn)
+ *     Name:myhost.default.service.arpa.
+ *     RecordData:[001900010000e02d00440201030d4983605c0406803deb2d672cc42224773977]
+ * Done
+ * @encode
```
```suggestion
 * @endcode
```

EskoDijk commented:
Reviewing currently - a few comments pending. I'll try to finish it next week!

EskoDijk commented on src/cli/README.md:
```diff
@@ -1561,6 +1561,24 @@ Service instance label is provided first, followed by the service name (note tha
+Requires `OPENTHREAD_CONFIG_DNS_CLIENT_ENABLE` and `OPENTHREAD_CONFIG_DNS_CLIENT_ARBITRARY_RECORD_QUERY_ENABLE`.
+
+Sends a DNS query for a given record type and DNS name. DNS name is provided as a first label, followed by the next labels which are dot '.' separated. Note that the first label can itself contain the dot '.' character.
+
```
Can add a sentence here that the <record-type> is a uint16 encoding one of the DNS RRTYPEs.

abtink replied:
Done! Thanks.

EskoDijk commented on src/core/config/dns_client.h:
```diff
@@ -84,6 +84,15 @@
+/**
+ * @def OPENTHREAD_CONFIG_DNS_CLIENT_ARBITRARY_RECORD_QUERY_ENABLE
+ *
+ * Define to 1 to enable support query or arbitrary record type.
```
```suggestion
 * Define to 1 to enable support for a query of arbitrary DNS record type.
```

EskoDijk commented on src/core/net/dns_client.cpp:
```diff
@@ -374,6 +374,49 @@ Error Client::Response::ReadTxtRecord(Section aSection, const Name &aName, Servi
+    uint16_t       offset;
+    ResourceRecord record;
+
+    VerifyOrExit(aIndex < mAnswerRecordCount + mAuthorityRecordCount + mAdditionalRecordCount, error = kErrorNotFound);
```
so the Question records are not included in the indexed records, right?

abtink replied:
No. Not incldued.

EskoDijk commented on src/core/net/dns_client.hpp:
```diff
@@ -560,6 +586,62 @@ class Client : public InstanceLocator, private NonCopyable
+
+    public:
+        /**
+         * Gets the query name associated with a record DNS query..
```
```suggestion
         * Gets the query name associated with a record DNS query.
```

EskoDijk commented on src/core/net/dns_client.hpp:
```diff
@@ -724,6 +806,33 @@ class Client : public InstanceLocator, private NonCopyable
+     * the config for this query. In a non-NULL @p aConfig, some of the fields can be left unspecified (value zero).
+     * The unspecified fields are then replaced by the values from the default config.
+     *
+     * @param[in] aRecordType   The resource record type to query.
```
```suggestion
     * @param[in] aRecordType   The resource record type (RRType) to query.
```

EskoDijk commented on src/core/net/dns_client.hpp:
```diff
@@ -724,6 +806,33 @@ class Client : public InstanceLocator, private NonCopyable
+     * @param[in] aConfig       A pointer to the config to use for this query.
+     *
+     * @retval kErrorNone          Query sent successfully. @p aCallback will be invoked to report the outcome.
+     * @retval kErrorNotBufs       Insufficient buffer to prepare and send query.
```
```suggestion
     * @retval kErrorNoBufs        Insufficient buffer to prepare and send query.
```

EskoDijk replied:
Note there was one other instance of `kErrorNotBufs` in file dns_client.cpp  (existing code)

abtink replied:
Thanks. Both fixed.

EskoDijk commented on src/core/net/dns_client.hpp:
```diff
@@ -724,6 +806,33 @@ class Client : public InstanceLocator, private NonCopyable
+     * @param[in] aNextLabels   The next labels of the name to be queried (MUST NOT be NULL).
+     * @param[in] aCallback     A function pointer that shall be called on response reception or time-out.
+     * @param[in] aContext      A pointer to arbitrary context information used with @p aCallback.
+     * @param[in] aConfig       A pointer to the config to use for this query.
```
For completeness here, should we add:

```suggestion
     * @param[in] aConfig       A pointer to the config to use for this query (or NULL for default config).
```

abtink commented on include/openthread/dns_client.h:
```diff
@@ -646,6 +646,118 @@ otError otDnsServiceResponseGetHostAddress(const otDnsServiceResponse *aResponse
+    uint8_t *mDataBuffer;      ///< Buffer to output the record data (MUST NOT be NULL).
+    uint16_t mDataBufferSize;  ///< On input, size of `mDataBuffer`. On output number of bytes written.
+    bool     mInAnswerSection; ///< Indicates whether the record is in Answer section.
+    bool     mInAuthSection;   ///< Indicates whether the record is in Authority section.
```
Yes.

abtink replied:
In the new push I updated this to be `mSection` with it value from an `enum`.

abtink commented on src/cli/cli_dns.cpp:
```diff
@@ -650,6 +699,55 @@ void Dns::HandleDnsServiceResponse(otError aError, const otDnsServiceResponse *a
+        recordInfo.mDataBuffer     = data;
+        recordInfo.mDataBufferSize = sizeof(name);
+
+        aError = otDnsRecordResponseGetRecordInfo(aResponse, index, &recordInfo);
```
The entire `otDnsRecordResponseGetRecordInfo()` and `otDnsRecordInfo` is now updated to allow partial record data reads (i.e., if it does not fit in the provided buffer).

Please check the updated API documentation.

abtink commented:
> Some issues raised. Also, should there be test code for the new CLI command? If this is enabled by default then maybe good to have a test also.

Thansk @EskoDijk. 

Regarding tests, they are all included in #11357 (as mentioned in PR desicption above). I added new tests specifically for CLI output (`test_dnssd.py`)

There is also another big change that I added in this PR.

Certain record types (SRV, PTR, CNAME) can have a DNS name in their record data, which may use DNS name compression in the response message.

I added new code to ensure they provide uncompressed data (along with detailed test to cover this in #11357).

Please check the code and updated API documentation.

```c++
/**
 * Reads the records from a DNS query response.
 *
 * MUST only be used from `otDnsRecordCallback`.
 *
 * The response may include multiple records. @p aIndex can be used to iterate through the list. Index zero gets the
 * first record and so on. When we reach the end of the list, `OT_ERROR_NOT_FOUND` is returned.
 *
 * Upon successful retrieval (`OT_ERROR_NONE`):
 * - `mRecordLength` is set to the actual length of the record's data.
 * - The data is copied into `mDataBuffer` (if not `NULL`) up to its capacity specified by `mDataBufferSize`.
 * - `mDataBufferSize` is then updated to reflect the number of bytes actually written into `mDataBuffer`.
 *
 * If the retrieved record type is PTR (12), CNAME (5), or SRV (33), the record data in the received response contains
 * a DNS name which may use DNS name compression. For these specific record types, the record data is first
 * decompressed such that it contains the full uncompressed DNS name. This decompressed data is then provided in
 * `mDataBuffer`, and `mRecordDataLength` will indicate the length of this decompressed data. For all other record
 * types, the record data is read and provided as it appears in the received response message.
 *
 * @param[in]  aResponse       A pointer to the response.
 * @param[in]  aIndex          The record index to retrieve.
 * @param[out] aRecordInfo     A pointer to a `otDnsRecordInfo` to populate with read record info.
 *
 * @retval OT_ERROR_NONE       The record was read successfully.
 * @retval OT_ERROR_NOT_FOUND  No record in @p aResponse at @p aIndex.
 * @retval OT_ERROR_PARSE      Could not parse the records in the @p aResponse.
 * @retval OT_ERROR_NO_BUFS    The record name does not fit in the provided `mNameBufferSize` in @p aRecordInfo, or
 *                             failed to allocate buffer to decompress a compressed DNS name (PTR, SRV, CNAME).
 */
otError otDnsRecordResponseGetRecordInfo(const otDnsRecordResponse *aResponse,
                                         uint16_t                   aIndex,
                                         otDnsRecordInfo           *aRecordInfo);
```

abtink commented:
@EskoDijk, I previously added a mechanism to uncompress DNS names in retrieved CNAME, PTR, and SRV record data (on the client when the query API is used). I have now added support for NS and DNAME records (since their RDATA format is simple, making support easy to add). This is related to the discussion in [https://github.com/openthread/openthread/pull/11357#discussion_r2031505617](https://github.com/openthread/openthread/pull/11357#discussion_r2031505617).

EskoDijk commented on src/cli/cli_dns.cpp:
```diff
@@ -650,6 +699,78 @@ void Dns::HandleDnsServiceResponse(otError aError, const otDnsServiceResponse *a
+    const char *const kSectionString[] = {
+        "answer",          // (0) OT_DNS_SECTION_ANSWER
+        "authority",       // (1) OT_DNS_SECTION_AUTHORITY
+        "additional data", // (2) OT_DNS_SECTION_ADDITIONAL_DATA
```
RFC 1035 uses "additional section" or "additional records" section. Follow-up DNS RFCs use one of these or "additional data" section. So it's very inconsistent overall. Given that, I would use only "additional" here.

```suggestion
        "additional", // (2) OT_DNS_SECTION_ADDITIONAL_DATA
```

EskoDijk replied:
Also remove the `_DATA` from the Additional section name, I think. (For consistency - it should be clear enough.)

EskoDijk commented on src/core/net/dns_client.hpp:
```diff
@@ -724,6 +831,33 @@ class Client : public InstanceLocator, private NonCopyable
+     * Sends a DNS query for a given record type and name.
+     *
+     * The @p aConfig can be `nullptr`. In this case the default config (from `GetDefaultConfig()`) will be used as
+     * the config for this query. In a non-NULL @p aConfig, some of the fields can be left unspecified (value zero).
```
It says "in a non-NULL aConfig" - but if aConfig is NULL, then the fields left unspecified will be taken from the default config, right?  So it doesn't matter if aConfig is NULL or not?

abtink replied:
Yes, exactly. If it is `NULL`, everything (in the config) is effectively unspecified, and therefore the specified default configuration values will be used.

The description elaborates a bit more on the `NULL` case to make it clear that the full `GetDefaultConfig()` values will be used.

Vyrastas commented on include/openthread/dns_client.h:
```diff
@@ -646,6 +646,139 @@ otError otDnsServiceResponseGetHostAddress(const otDnsServiceResponse *aResponse
+ * The @p aError can have the following:
+ *
+ *  - OT_ERROR_NONE              A response was received successfully.
+ *  - OT_ERROR_ABORT             A DNS transaction was aborted by stack.
```
```suggestion
 *  - OT_ERROR_ABORT             A DNS transaction was aborted by the stack.
```

Vyrastas commented on src/core/net/dns_client.hpp:
```diff
@@ -560,6 +598,75 @@ class Client : public InstanceLocator, private NonCopyable
+        /**
+         * Gets the query name associated with a record DNS query.
+         *
+         * MUST only be used from `RecordCallback`
```
```suggestion
         * MUST only be used from `RecordCallback`.
```

Vyrastas commented on src/core/net/dns_client.hpp:
```diff
@@ -560,6 +598,75 @@ class Client : public InstanceLocator, private NonCopyable
+         * The response may include multiple records. @p aIndex can be used to iterate through the list. Index zero gets
+         * the first record and so on. When we reach the end of the list, `kErrorNotFound` is returned.
+         *
+         * Upon successful retrieval (`kErrorNone`)
```
```suggestion
         * Upon successful retrieval (`kErrorNone`):
```

Vyrastas commented on src/core/net/dns_types.hpp:
```diff
@@ -1267,6 +1267,7 @@ class ResourceRecord
     // Resource Record Types.
     static constexpr uint16_t kTypeZero  = 0;   ///< Zero as special indicator for the SIG RR (SIG(0) from RFC 2931).
     static constexpr uint16_t kTypeA     = 1;   ///< Address record (IPv4).
+    static constexpr uint16_t kTypeNs    = 2;   ///< NS record (an authoritative name server)
```
```suggestion
    static constexpr uint16_t kTypeNs    = 2;   ///< NS record (an authoritative name server).
```

abtink replied:
Thanks. All fixed in the new pushed commit.

Vyrastas reviewed (changes requested):
Just a few minor nits after reviewing again
